service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6260,419 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+		&current.WorkflowTaskScheduledTime,
+		&current.WorkflowTaskOriginalScheduledTime,
+		&current.WorkflowTaskType,
+		&current.WorkflowTaskSuggestContinueAsNew,
+		&current.WorkflowTaskHistorySizeBytes,
+		&current.WorkflowTaskBuildId,
+		&current.WorkflowTaskBuildIdRedirectCounter,
+		&current.VersionHistories,
+		&current.ExecutionStats,
+		&current.LastFirstEventTxnId,
+		&current.ParentClock,
+		&current.CloseTransferTaskId,
+		&current.CloseVisibilityTaskId,
+		&current.CloseVisibilityTaskCompleted,
+		&current.WorkflowExecutionTimerTaskStatus,
+		&current.SubStateMachinesByType,
+		&current.StateMachineTimers,
+		&current.TaskGenerationShardClockTimestamp,
+		&current.UpdateInfos,
+	}
+	err := common.MergeProtoExcludingFields(current, incoming, doNotSyncFields...)
+	if err != nil {
+		return err
+	}
+
+	ms.ClearStickyTaskQueue()
+
+	if ms.executionInfo.WorkflowTaskType == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		ms.workflowTaskManager.deleteWorkflowTask()
+	} else {
+		taskInfo := &WorkflowTaskInfo{
+			Version:             incoming.WorkflowTaskVersion,
+			ScheduledEventID:    incoming.WorkflowTaskScheduledEventId,
+			StartedEventID:      incoming.WorkflowTaskStartedEventId,
+			RequestID:           incoming.WorkflowTaskRequestId,
+			WorkflowTaskTimeout: incoming.WorkflowTaskTimeout.AsDuration(),
+			Attempt:             incoming.WorkflowTaskAttempt,
+			StartedTime:         incoming.WorkflowTaskStartedTime.AsTime(),
+			ScheduledTime:       incoming.WorkflowTaskScheduledTime.AsTime(),
+
+			OriginalScheduledTime: incoming.WorkflowTaskOriginalScheduledTime.AsTime(),
+			Type:                  incoming.WorkflowTaskType,
+
+			SuggestContinueAsNew:   incoming.WorkflowTaskSuggestContinueAsNew,
+			HistorySizeBytes:       incoming.WorkflowTaskHistorySizeBytes,
+			BuildId:                incoming.WorkflowTaskBuildId,
+			BuildIdRedirectCounter: incoming.WorkflowTaskBuildIdRedirectCounter,
+		}
+		ms.workflowTaskManager.UpdateWorkflowTask(taskInfo)
+	}
+
+	return nil
+}
+
+func (ms *MutableStateImpl) syncSubStateMachinesByType(incoming map[string]*persistencespb.StateMachineMap) error {
+	currentHSM := ms.HSM()
+
+	// we don't care about the root here which is the entire mutable state
+	incomingHSM, err := hsm.NewRoot(
+		ms.shard.StateMachineRegistry(),
+		StateMachineType,
+		ms,
+		incoming,
+		ms,
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := incomingHSM.Walk(func(incomingNode *hsm.Node) error {
+		if incomingNode.Parent == nil {
+			// skip root which is the entire mutable state
+			return nil
+		}
+
+		incomingNodePath := incomingNode.Path()
+		currentNode, err := currentHSM.Child(incomingNodePath)
+		if err != nil {
+			// 1. Already done history resend if needed before,
+			// and node creation today always associated with an event
+			// 2. Node deletion is not supported right now.
+			// Based on 1 and 2, node should always be found here.
+			return err
+		}
+
+		return currentNode.Sync(incomingNode)
```

Comment:
This internally performs a task refresh for the synced node. If we go with this approach (which I am fine with), caller needs to clear all generated tasks (can be done with ms.PopTasks()) before doing a task refresh. cc @xwduan